File: Shared/shared-types/src/easing.ts
/**
 * Easing — one specification, two implementations.
 *
 * The twin is `services/render-engine/src/easing.rs`. They are one specification in two
 * languages, exactly like `resolveTextureFit`/`resolve_texture_fit` and the tile system: the
 * Editor previews with this file and Program renders with that one, so a curve that differs by
 * a thousandth is a graphic that eases differently on air than it did in the viewport. Both
 * sides are checked against `Shared/animation-engine/fixtures/easing-vectors.json`, which is
 * generated once and never regenerated to make a test pass.
 *
 * ## Why closed forms rather than a table
 *
 * Every curve here is a closed form or a fixed-iteration solve, written with the same
 * operations in the same order in both languages. IEEE 754 doubles then agree to well inside
 * the 1e-9 the conformance test demands. A lookup table would have to agree on its sampling
 * too, and would be wrong between samples.
 *
 * ## Why this lives in `shared-types` and not `animation-engine`
 *
 * `animation-engine` depends on `shared-types`, and the evaluator that consumes easing
 * (`sampleChannel`, `evaluateSceneAtFrame`) lives here. Putting the implementation in
 * `animation-engine` would force this package to keep its own copy — a second TypeScript
 * easing implementation, which is the drift this phase exists to remove. `@grapix/animation-engine`
 * re-exports it, so the package boundary reads the way the plan intended.
 *
 * ## Compatibility
 *
 * `ease-in`, `ease-out` and `ease-in-out` are the three curves scenes on disk already use, and
 * they are quadratic. They keep that exact shape and are aliases of the `-quad` forms. Adding a
 * name here never changes a curve that already existed.
 */

/**
 * Every easing a keyframe may name.
 *
 * Kebab-case, matching the three values already written into scene documents. The bare
 * `ease-in`/`ease-out`/`ease-in-out` are quadratic for compatibility; `ease` is the CSS default
 * curve, which is not the same thing.
 */
export type SceneKeyframeEasing =
  | "linear"
  | "hold"
  | "ease"
  | "ease-in"
  | "ease-out"
  | "ease-in-out"
  | "ease-in-quad" | "ease-out-quad" | "ease-in-out-quad"
  | "ease-in-cubic" | "ease-out-cubic" | "ease-in-out-cubic"
  | "ease-in-quart" | "ease-out-quart" | "ease-in-out-quart"
  | "ease-in-quint" | "ease-out-quint" | "ease-in-out-quint"
  | "ease-in-sine" | "ease-out-sine" | "ease-in-out-sine"
  | "ease-in-expo" | "ease-out-expo" | "ease-in-out-expo"
  | "ease-in-circ" | "ease-out-circ" | "ease-in-out-circ"
  | "ease-in-back" | "ease-out-back" | "ease-in-out-back"
  | "ease-in-elastic" | "ease-out-elastic" | "ease-in-out-elastic"
  | "ease-in-bounce" | "ease-out-bounce" | "ease-in-out-bounce";

/** Every easing name, in the order the fixture and any picker should list them. */
export const SCENE_KEYFRAME_EASINGS: readonly SceneKeyframeEasing[] = [
  "linear",
  "hold",
  "ease",
  "ease-in",
  "ease-out",
  "ease-in-out",
  "ease-in-quad", "ease-out-quad", "ease-in-out-quad",
  "ease-in-cubic", "ease-out-cubic", "ease-in-out-cubic",
  "ease-in-quart", "ease-out-quart", "ease-in-out-quart",
  "ease-in-quint", "ease-out-quint", "ease-in-out-quint",
  "ease-in-sine", "ease-out-sine", "ease-in-out-sine",
  "ease-in-expo", "ease-out-expo", "ease-in-out-expo",
  "ease-in-circ", "ease-out-circ", "ease-in-out-circ",
  "ease-in-back", "ease-out-back", "ease-in-out-back",
  "ease-in-elastic", "ease-out-elastic", "ease-in-out-elastic",
  "ease-in-bounce", "ease-out-bounce", "ease-in-out-bounce"
];

export function isSceneKeyframeEasing(value: unknown): value is SceneKeyframeEasing {
  return typeof value === "string" && (SCENE_KEYFRAME_EASINGS as readonly string[]).includes(value);
}

// --- constants, named so the Rust twin can be compared line by line ---------

/** `back` overshoot, the classic Penner constant. */
const BACK_C1 = 1.70158;
const BACK_C3 = BACK_C1 + 1;
const BACK_C2 = BACK_C1 * 1.525;
/** `elastic` angular frequencies. */
const ELASTIC_C4 = (2 * Math.PI) / 3;
const ELASTIC_C5 = (2 * Math.PI) / 4.5;
/** `bounce` gravity constant and segment width. */
const BOUNCE_N1 = 7.5625;
const BOUNCE_D1 = 2.75;

/**
 * Solve a CSS-style cubic-bezier for `y` at `x`.
 *
 * A fixed iteration count with no early exit, deliberately: an epsilon-based break can take a
 * different number of steps in two languages and land a fraction apart, and this has to match
 * the Rust twin to 1e-9. Eight Newton steps from `x` as the initial guess is well converged for
 * every curve declared here; the derivative guard keeps a flat region from dividing by zero.
 */
function cubicBezier(x1: number, y1: number, x2: number, y2: number, x: number): number {
  const curveX = (t: number) => {
    const u = 1 - t;
    return 3 * u * u * t * x1 + 3 * u * t * t * x2 + t * t * t;
  };
  const curveY = (t: number) => {
    const u = 1 - t;
    return 3 * u * u * t * y1 + 3 * u * t * t * y2 + t * t * t;
  };
  const slopeX = (t: number) => {
    const u = 1 - t;
    return 3 * u * u * (x1 - 0) + 6 * u * t * (x2 - x1) + 3 * t * t * (1 - x2);
  };

  let t = x;
  for (let step = 0; step < 8; step += 1) {
    const dx = curveX(t) - x;
    const d = slopeX(t);
    if (Math.abs(d) < 1e-12) break;
    t -= dx / d;
  }
  return curveY(t < 0 ? 0 : t > 1 ? 1 : t);
}

/**
 * Evaluate an easing at `t`.
 *
 * `t` is clamped to [0,1] because a caller that computed a position outside the segment has
 * already made its mistake; extrapolating an elastic curve past 1 produces a value nobody
 * authored. An unrecognised name returns `undefined` rather than falling back to linear — the
 * caller decides, and `sampleChannel` holds the previous value and reports it.
 */
export function applyEasing(easing: string, t: number): number | undefined {
  const c = t < 0 ? 0 : t > 1 ? 1 : t;
  switch (easing) {
    case "linear":
      return c;
    // A hold key keeps its value for the whole segment and steps at the next key. Expressed as
    // an easing so every path through the evaluator treats it the same way.
    case "hold":
      return c < 1 ? 0 : 1;
    case "ease":
      return cubicBezier(0.25, 0.1, 0.25, 1, c);

    case "ease-in":
    case "ease-in-quad":
      return c * c;
    case "ease-out":
    case "ease-out-quad":
      return 1 - (1 - c) * (1 - c);
    case "ease-in-out":
    case "ease-in-out-quad":
      return c < 0.5 ? 2 * c * c : 1 - Math.pow(-2 * c + 2, 2) / 2;

    case "ease-in-cubic":
      return c * c * c;
    case "ease-out-cubic":
      return 1 - Math.pow(1 - c, 3);
    case "ease-in-out-cubic":
      return c < 0.5 ? 4 * c * c * c : 1 - Math.pow(-2 * c + 2, 3) / 2;

    case "ease-in-quart":
      return c * c * c * c;
    case "ease-out-quart":
      return 1 - Math.pow(1 - c, 4);
    case "ease-in-out-quart":
      return c < 0.5 ? 8 * c * c * c * c : 1 - Math.pow(-2 * c + 2, 4) / 2;

    case "ease-in-quint":
      return c * c * c * c * c;
    case "ease-out-quint":
      return 1 - Math.pow(1 - c, 5);
    case "ease-in-out-quint":
      return c < 0.5 ? 16 * c * c * c * c * c : 1 - Math.pow(-2 * c + 2, 5) / 2;

    case "ease-in-sine":
      return 1 - Math.cos((c * Math.PI) / 2);
    case "ease-out-sine":
      return Math.sin((c * Math.PI) / 2);
    case "ease-in-out-sine":
      return -(Math.cos(Math.PI * c) - 1) / 2;

    case "ease-in-expo":
      return c === 0 ? 0 : Math.pow(2, 10 * c - 10);
    case "ease-out-expo":
      return c === 1 ? 1 : 1 - Math.pow(2, -10 * c);
    case "ease-in-out-expo":
      if (c === 0) return 0;
      if (c === 1) return 1;
      return c < 0.5 ? Math.pow(2, 20 * c - 10) / 2 : (2 - Math.pow(2, -20 * c + 10)) / 2;

    case "ease-in-circ":
      return 1 - Math.sqrt(1 - Math.pow(c, 2));
    case "ease-out-circ":
      return Math.sqrt(1 - Math.pow(c - 1, 2));
    case "ease-in-out-circ":
      return c < 0.5
        ? (1 - Math.sqrt(1 - Math.pow(2 * c, 2))) / 2
        : (Math.sqrt(1 - Math.pow(-2 * c + 2, 2)) + 1) / 2;

    case "ease-in-back":
      return BACK_C3 * c * c * c - BACK_C1 * c * c;
    case "ease-out-back":
      return 1 + BACK_C3 * Math.pow(c - 1, 3) + BACK_C1 * Math.pow(c - 1, 2);
    case "ease-in-out-back":
      return c < 0.5
        ? (Math.pow(2 * c, 2) * ((BACK_C2 + 1) * 2 * c - BACK_C2)) / 2
        : (Math.pow(2 * c - 2, 2) * ((BACK_C2 + 1) * (c * 2 - 2) + BACK_C2) + 2) / 2;

    case "ease-in-elastic":
      if (c === 0) return 0;
      if (c === 1) return 1;
      return -Math.pow(2, 10 * c - 10) * Math.sin((c * 10 - 10.75) * ELASTIC_C4);
    case "ease-out-elastic":
      if (c === 0) return 0;
      if (c === 1) return 1;
      return Math.pow(2, -10 * c) * Math.sin((c * 10 - 0.75) * ELASTIC_C4) + 1;
    case "ease-in-out-elastic":
      if (c === 0) return 0;
      if (c === 1) return 1;
      return c < 0.5
        ? -(Math.pow(2, 20 * c - 10) * Math.sin((20 * c - 11.125) * ELASTIC_C5)) / 2
        : (Math.pow(2, -20 * c + 10) * Math.sin((20 * c - 11.125) * ELASTIC_C5)) / 2 + 1;

    case "ease-in-bounce":
      return 1 - bounceOut(1 - c);
    case "ease-out-bounce":
      return bounceOut(c);
    case "ease-in-out-bounce":
      return c < 0.5 ? (1 - bounceOut(1 - 2 * c)) / 2 : (1 + bounceOut(2 * c - 1)) / 2;

    default:
      return undefined;
  }
}

function bounceOut(t: number): number {
  if (t < 1 / BOUNCE_D1) return BOUNCE_N1 * t * t;
  if (t < 2 / BOUNCE_D1) {
    const shifted = t - 1.5 / BOUNCE_D1;
    return BOUNCE_N1 * shifted * shifted + 0.75;
  }
  if (t < 2.5 / BOUNCE_D1) {
    const shifted = t - 2.25 / BOUNCE_D1;
    return BOUNCE_N1 * shifted * shifted + 0.9375;
  }
  const shifted = t - 2.625 / BOUNCE_D1;
  return BOUNCE_N1 * shifted * shifted + 0.984375;
}

// --- diagnostics ------------------------------------------------------------

/**
 * A fault the evaluator found while sampling.
 *
 * Reported through a sink rather than thrown or logged: this package has no dependencies and
 * runs in a browser, an installer-bundled service and a test alike, so it cannot know what a
 * diagnostics log is. The Editor and the control service install a listener that forwards to
 * their own console; a host that installs none loses nothing but the message.
 */
export interface AnimationDiagnostic {
  code: "animation.unknown-easing";
  message: string;
  /** The offending value, verbatim, so the report can name what was written. */
  value: string;
  frame: number;
  objectId?: string;
  property?: string;
}

const animationDiagnosticListeners = new Set<(diagnostic: AnimationDiagnostic) => void>();

export function onAnimationDiagnostic(listener: (diagnostic: AnimationDiagnostic) => void): () => void {
  animationDiagnosticListeners.add(listener);
  return () => animationDiagnosticListeners.delete(listener);
}

/**
 * Faults already reported, so a per-frame sampler cannot flood the console.
 *
 * The evaluator runs this property every time the scene is sampled — fifty times a second
 * during playback. One real fault would then push hundreds of identical records a second and
 * push everything else out of a bounded console, which turns a useful report into a denial of
 * service against itself. A fault is therefore reported once per object, property and offending
 * value; `resetAnimationDiagnostics` clears the memo when a host loads a different scene.
 */
const reportedAnimationFaults = new Set<string>();

export function resetAnimationDiagnostics(): void {
  reportedAnimationFaults.clear();
}

/**
 * Report a fault once, never throwing.
 *
 * A listener that throws — or one that is slow — must not be able to stop a scene evaluating.
 * The caller has already decided what to render.
 */
export function reportAnimationDiagnostic(diagnostic: AnimationDiagnostic): void {
  const fault = `${diagnostic.code}|${diagnostic.objectId ?? ""}|${diagnostic.property ?? ""}|${diagnostic.value}`;
  if (reportedAnimationFaults.has(fault)) return;
  reportedAnimationFaults.add(fault);

  for (const listener of animationDiagnosticListeners) {
    try {
      listener(diagnostic);
    } catch {
      // A diagnostic that fails to deliver must not become the fault it was reporting.
    }
  }
}

File: services/render-engine/src/easing.rs
//! Easing — the Rust half of one specification.
//!
//! The twin is `Shared/shared-types/src/easing.ts`. The Editor previews with that one and
//! Program renders with this one, so they are one specification in two languages, exactly like
//! `resolve_texture_fit` and the tile system. Both are checked against
//! `Shared/animation-engine/fixtures/easing-vectors.json`, which is generated once and never
//! regenerated to make a test pass: a changed number there is a changed curve, and every scene
//! already authored against it would then animate differently on air.
//!
//! Every curve is a closed form or a fixed-iteration solve, written with the same operations in
//! the same order as the TypeScript. IEEE 754 doubles then agree far inside the 1e-9 the
//! conformance test demands — which is why there is no lookup table and no early exit anywhere
//! in this file.
//!
//! `ease-in`, `ease-out` and `ease-in-out` are the three curves scenes on disk already use and
//! they are quadratic; they are aliases of the `-quad` forms and their shape is frozen.

/// `back` overshoot, the classic Penner constant.
const BACK_C1: f64 = 1.70158;
const BACK_C3: f64 = BACK_C1 + 1.0;
const BACK_C2: f64 = BACK_C1 * 1.525;
/// `elastic` angular frequencies.
const ELASTIC_C4: f64 = (2.0 * std::f64::consts::PI) / 3.0;
const ELASTIC_C5: f64 = (2.0 * std::f64::consts::PI) / 4.5;
/// `bounce` gravity constant and segment width.
const BOUNCE_N1: f64 = 7.5625;
const BOUNCE_D1: f64 = 2.75;

/// Every easing name this build implements, in the fixture's order.
pub const SCENE_KEYFRAME_EASINGS: [&str; 36] = [
    "linear",
    "hold",
    "ease",
    "ease-in",
    "ease-out",
    "ease-in-out",
    "ease-in-quad",
    "ease-out-quad",
    "ease-in-out-quad",
    "ease-in-cubic",
    "ease-out-cubic",
    "ease-in-out-cubic",
    "ease-in-quart",
    "ease-out-quart",
    "ease-in-out-quart",
    "ease-in-quint",
    "ease-out-quint",
    "ease-in-out-quint",
    "ease-in-sine",
    "ease-out-sine",
    "ease-in-out-sine",
    "ease-in-expo",
    "ease-out-expo",
    "ease-in-out-expo",
    "ease-in-circ",
    "ease-out-circ",
    "ease-in-out-circ",
    "ease-in-back",
    "ease-out-back",
    "ease-in-out-back",
    "ease-in-elastic",
    "ease-out-elastic",
    "ease-in-out-elastic",
    "ease-in-bounce",
    "ease-out-bounce",
    "ease-in-out-bounce",
];

/// Solve a CSS-style cubic-bezier for `y` at `x`.
///
/// A fixed iteration count with no early exit, matching the TypeScript exactly: an
/// epsilon-based break can take a different number of steps in two languages and land a
/// fraction apart, and this has to agree to 1e-9.
fn cubic_bezier(x1: f64, y1: f64, x2: f64, y2: f64, x: f64) -> f64 {
    let curve_x = |t: f64| {
        let u = 1.0 - t;
        3.0 * u * u * t * x1 + 3.0 * u * t * t * x2 + t * t * t
    };
    let curve_y = |t: f64| {
        let u = 1.0 - t;
        3.0 * u * u * t * y1 + 3.0 * u * t * t * y2 + t * t * t
    };
    let slope_x = |t: f64| {
        let u = 1.0 - t;
        3.0 * u * u * (x1 - 0.0) + 6.0 * u * t * (x2 - x1) + 3.0 * t * t * (1.0 - x2)
    };

    let mut t = x;
    for _ in 0..8 {
        let dx = curve_x(t) - x;
        let d = slope_x(t);
        if d.abs() < 1e-12 {
            break;
        }
        t -= dx / d;
    }
    curve_y(t.clamp(0.0, 1.0))
}

fn bounce_out(t: f64) -> f64 {
    if t < 1.0 / BOUNCE_D1 {
        return BOUNCE_N1 * t * t;
    }
    if t < 2.0 / BOUNCE_D1 {
        let shifted = t - 1.5 / BOUNCE_D1;
        return BOUNCE_N1 * shifted * shifted + 0.75;
    }
    if t < 2.5 / BOUNCE_D1 {
        let shifted = t - 2.25 / BOUNCE_D1;
        return BOUNCE_N1 * shifted * shifted + 0.9375;
    }
    let shifted = t - 2.625 / BOUNCE_D1;
    BOUNCE_N1 * shifted * shifted + 0.984375
}

/// Evaluate an easing at `t`, or `None` when this build does not implement the name.
///
/// `None` rather than a linear fallback, deliberately and identically to the TypeScript: a
/// scene authored against an easing this build lacks must hold its previous value and raise a
/// diagnostic, not animate confidently along a curve nobody chose.
pub fn apply_easing(easing: &str, t: f64) -> Option<f64> {
    let c = t.clamp(0.0, 1.0);
    let value = match easing {
        "linear" => c,
        // A hold key keeps its value for the whole segment and steps at the next key.
        "hold" => {
            if c < 1.0 {
                0.0
            } else {
                1.0
            }
        }
        "ease" => cubic_bezier(0.25, 0.1, 0.25, 1.0, c),

        "ease-in" | "ease-in-quad" => c * c,
        "ease-out" | "ease-out-quad" => 1.0 - (1.0 - c) * (1.0 - c),
        "ease-in-out" | "ease-in-out-quad" => {
            if c < 0.5 {
                2.0 * c * c
            } else {
                1.0 - (-2.0 * c + 2.0).powi(2) / 2.0
            }
        }

        "ease-in-cubic" => c * c * c,
        "ease-out-cubic" => 1.0 - (1.0 - c).powi(3),
        "ease-in-out-cubic" => {
            if c < 0.5 {
                4.0 * c * c * c
            } else {
                1.0 - (-2.0 * c + 2.0).powi(3) / 2.0
            }
        }

        "ease-in-quart" => c * c * c * c,
        "ease-out-quart" => 1.0 - (1.0 - c).powi(4),
        "ease-in-out-quart" => {
            if c < 0.5 {
                8.0 * c * c * c * c
            } else {
                1.0 - (-2.0 * c + 2.0).powi(4) / 2.0
            }
        }

        "ease-in-quint" => c * c * c * c * c,
        "ease-out-quint" => 1.0 - (1.0 - c).powi(5),
        "ease-in-out-quint" => {
            if c < 0.5 {
                16.0 * c * c * c * c * c
            } else {
                1.0 - (-2.0 * c + 2.0).powi(5) / 2.0
            }
        }

        "ease-in-sine" => 1.0 - ((c * std::f64::consts::PI) / 2.0).cos(),
        "ease-out-sine" => ((c * std::f64::consts::PI) / 2.0).sin(),
        "ease-in-out-sine" => -((std::f64::consts::PI * c).cos() - 1.0) / 2.0,

        "ease-in-expo" => {
            if c == 0.0 {
                0.0
            } else {
                (2.0_f64).powf(10.0 * c - 10.0)
            }
        }
        "ease-out-expo" => {
            if c == 1.0 {
                1.0
            } else {
                1.0 - (2.0_f64).powf(-10.0 * c)
            }
        }
        "ease-in-out-expo" => {
            if c == 0.0 {
                0.0
            } else if c == 1.0 {
                1.0
            } else if c < 0.5 {
                (2.0_f64).powf(20.0 * c - 10.0) / 2.0
            } else {
                (2.0 - (2.0_f64).powf(-20.0 * c + 10.0)) / 2.0
            }
        }

        "ease-in-circ" => 1.0 - (1.0 - c.powi(2)).sqrt(),
        "ease-out-circ" => (1.0 - (c - 1.0).powi(2)).sqrt(),
        "ease-in-out-circ" => {
            if c < 0.5 {
                (1.0 - (1.0 - (2.0 * c).powi(2)).sqrt()) / 2.0
            } else {
                ((1.0 - (-2.0 * c + 2.0).powi(2)).sqrt() + 1.0) / 2.0
            }
        }

        "ease-in-back" => BACK_C3 * c * c * c - BACK_C1 * c * c,
        "ease-out-back" => 1.0 + BACK_C3 * (c - 1.0).powi(3) + BACK_C1 * (c - 1.0).powi(2),
        "ease-in-out-back" => {
            if c < 0.5 {
                ((2.0 * c).powi(2) * ((BACK_C2 + 1.0) * 2.0 * c - BACK_C2)) / 2.0
            } else {
                ((2.0 * c - 2.0).powi(2) * ((BACK_C2 + 1.0) * (c * 2.0 - 2.0) + BACK_C2) + 2.0) / 2.0
            }
        }

        "ease-in-elastic" => {
            if c == 0.0 {
                0.0
            } else if c == 1.0 {
                1.0
            } else {
                -(2.0_f64).powf(10.0 * c - 10.0) * ((c * 10.0 - 10.75) * ELASTIC_C4).sin()
            }
        }
        "ease-out-elastic" => {
            if c == 0.0 {
                0.0
            } else if c == 1.0 {
                1.0
            } else {
                (2.0_f64).powf(-10.0 * c) * ((c * 10.0 - 0.75) * ELASTIC_C4).sin() + 1.0
            }
        }
        "ease-in-out-elastic" => {
            if c == 0.0 {
                0.0
            } else if c == 1.0 {
                1.0
            } else if c < 0.5 {
                -((2.0_f64).powf(20.0 * c - 10.0) * ((20.0 * c - 11.125) * ELASTIC_C5).sin()) / 2.0
            } else {
                ((2.0_f64).powf(-20.0 * c + 10.0) * ((20.0 * c - 11.125) * ELASTIC_C5).sin()) / 2.0
                    + 1.0
            }
        }

        "ease-in-bounce" => 1.0 - bounce_out(1.0 - c),
        "ease-out-bounce" => bounce_out(c),
        "ease-in-out-bounce" => {
            if c < 0.5 {
                (1.0 - bounce_out(1.0 - 2.0 * c)) / 2.0
            } else {
                (1.0 + bounce_out(2.0 * c - 1.0)) / 2.0
            }
        }

        _ => return None,
    };
    Some(value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    /// The conformance table both languages answer to.
    ///
    /// Read from the checked-in fixture rather than recomputed here, because a test that
    /// recomputes the expectation from the implementation proves only that the implementation
    /// equals itself. This is the whole point of P0: add an easing to one language and not the
    /// other, or change a curve, and this fails locally rather than on air.
    #[test]
    fn every_easing_matches_the_shared_conformance_vectors() {
        let path = concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../Shared/animation-engine/fixtures/easing-vectors.json"
        );
        let text = std::fs::read_to_string(path)
            .unwrap_or_else(|error| panic!("easing fixture unreadable at {path}: {error}"));
        let fixture: Value = serde_json::from_str(&text).expect("easing fixture is not JSON");

        let times: Vec<f64> = fixture["times"]
            .as_array()
            .expect("fixture.times")
            .iter()
            .map(|value| value.as_f64().expect("time is a number"))
            .collect();
        let tolerance = fixture["tolerance"].as_f64().unwrap_or(1e-9);
        let easings = fixture["easings"].as_object().expect("fixture.easings");

        for (name, samples) in easings {
            let expected: Vec<f64> = samples
                .as_array()
                .expect("samples array")
                .iter()
                .map(|value| value.as_f64().expect("sample is a number"))
                .collect();
            assert_eq!(
                expected.len(),
                times.len(),
                "{name} has {} samples for {} times",
                expected.len(),
                times.len()
            );

            for (index, time) in times.iter().enumerate() {
                let actual = apply_easing(name, *time).unwrap_or_else(|| {
                    panic!("the Rust build does not implement easing \"{name}\"")
                });
                let difference = (actual - expected[index]).abs();
                assert!(
                    difference <= tolerance,
                    "{name} at t={time}: Rust {actual} vs fixture {} (difference {difference} exceeds {tolerance})",
                    expected[index]
                );
            }
        }
    }

    /// Both languages must implement exactly the same set — no more, no less.
    #[test]
    fn the_implemented_set_matches_the_fixture_exactly() {
        let path = concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../Shared/animation-engine/fixtures/easing-vectors.json"
        );
        let fixture: Value =
            serde_json::from_str(&std::fs::read_to_string(path).expect("fixture")).expect("json");
        let easings = fixture["easings"].as_object().expect("fixture.easings");

        for name in SCENE_KEYFRAME_EASINGS {
            assert!(
                easings.contains_key(name),
                "{name} is implemented in Rust but absent from the shared fixture"
            );
        }
        for name in easings.keys() {
            assert!(
                SCENE_KEYFRAME_EASINGS.contains(&name.as_str()),
                "{name} is in the shared fixture but not implemented in Rust"
            );
        }
    }

    /// An unknown easing is `None`, never a silent linear substitution.
    #[test]
    fn an_unknown_easing_is_refused_rather_than_guessed() {
        assert_eq!(apply_easing("ease-in-out-quintic-ish", 0.5), None);
        assert_eq!(apply_easing("", 0.5), None);
        // The legacy three keep their exact quadratic shape.
        assert_eq!(apply_easing("ease-in", 0.5), Some(0.25));
        assert_eq!(apply_easing("ease-out", 0.5), Some(0.75));
        assert_eq!(apply_easing("ease-in-out", 0.25), Some(0.125));
    }
}
